core/src/main/scala/org/apache/spark/sql/streaming/SnappySinkCallback.scala

```diff
@@ -0,0 +1,184 @@
+
+  @Override
+  def createSink(
+                  sqlContext: SQLContext,
+                  parameters: Map[String, String],
+                  partitionColumns: Seq[String],
+                  outputMode: OutputMode): Sink = {
+    createSinkStateTableIfNotExist(sqlContext)
+    val cc = try {
+      Utils.classForName(parameters(SINK_CALLBACK)).newInstance()
+    } catch {
+      case _: NoSuchElementException => new DefaultSnappySinkCallback()
+    }
+
+    SnappyStoreSink(sqlContext.asInstanceOf[SnappyContext].snappySession, parameters,
+      cc.asInstanceOf[SnappySinkCallback])
+  }
+
+  private def createSinkStateTableIfNotExist(sqlContext: SQLContext) = {
+    sqlContext.asInstanceOf[SnappyContext].snappySession.sql(s"create table if not exists" +
+      s" $SINK_STATE_TABLE (" +
+      " stream_query_id varchar(200)," +
+      " batch_id long, " +
+      " PRIMARY KEY (stream_query_id)) using row options(DISKSTORE 'GFXD-DD-DISKSTORE')")
+  }
+
+  @Override
+  def shortName(): String = SNAPPY_SINK_NAME
+
+}
+
+case class SnappyStoreSink(snappySession: SnappySession,
+    parameters: Map[String, String], sinkCallback: SnappySinkCallback) extends Sink {
+
+  override def addBatch(batchId: Long, data: Dataset[Row]): Unit = {
+    val streamQueryId = parameters(STREAM_QUERY_ID).toUpperCase
+
+    val updated = snappySession.sql(s"update $SINK_STATE_TABLE " +
+      s"set batch_id=$batchId where stream_query_id='$streamQueryId' and batch_id != $batchId")
+      .collect()(0).getAs("count").asInstanceOf[Long]
+
+    // TODO: use JDBC connection here
+    var posDup = false
+
+    if (updated == 0) {
+      try {
+        snappySession.insert(SINK_STATE_TABLE, Row(streamQueryId, batchId))
+        posDup = false
+      }
+      catch {
+        case e: SQLException if e.getSQLState.equals("23505") => posDup = true
+        case exception: Throwable => throw exception
+      }
+    }
+
+    sinkCallback.process(snappySession, parameters, batchId, convert(data), posDup)
+  }
+
+  /**
+    * This conversion is necessary as Sink
+    * documentation disallows an operation on incoming dataframe.
+    * Otherwise it will break incremental planning of streaming dataframes.
+    * See http://apache-spark-developers-list.1001551.n3.nabble.com/
+    * Structured-Streaming-Sink-in-2-0-collect-foreach-restrictions-added-in-
+    * SPARK-16020-td18118.html
+    * for a detailed discussion.
+    */
+  def convert(ds: DataFrame): DataFrame = {
+    snappySession.internalCreateDataFrame(
+      ds.queryExecution.toRdd,
+      StructType(ds.schema.fields))
+  }
+}
+
+object DefaultSnappySinkCallback {
+  private val log = Logger.getLogger(classOf[DefaultSnappySinkCallback].getName)
+}
+
+import org.apache.spark.sql.snappy._
+
+class DefaultSnappySinkCallback extends SnappySinkCallback {
+  def process(snappySession: SnappySession, parameters: Map[String, String],
+              batchId: Long, df: Dataset[Row], posDup: Boolean) {
+
+    df.cache()
```

Comment:
Will suggest adding a df.count() immediately after to force cache materialization. Sometimes caching on first query materialization causes unexpected cache due to different kind of plan. Also first execution will have incorrect stats that can be expensive depending on the query, so having a count() is safer. Specifically check if the deleteFrom leads to a SortMergeJoin plan rather than a broadcast under some cases. If so, then count() is definitely required.